File: leetcode/199.binary-tree-right-side-view.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode | None) -> list[int]:
        res: list[int] = []
        level: list[TreeNode] = [root] if root else []

        while level:
            res.append(level[-1].val)
            next_level: list[TreeNode] = []

            for node in level:
                for child in (node.left, node.right):
                    if child:
                        next_level.append(child)

            level = next_level

        return res


# @leet end

File: leetcode/199.binary-tree-right-side-view.ts
// @leet start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
  const result: number[] = [];
  let curr: TreeNode[] = root ? [root] : [];

  while (curr.length) {
    result.push(curr[curr.length - 1].val);

    const next: TreeNode[] = [];

    for (const node of curr) {
      if (node.left) {
        next.push(node.left);
      }

      if (node.right) {
        next.push(node.right);
      }
    }

    curr = next;
  }

  return result;
}
// @leet end
